snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -0,0 +1,150 @@
+from collections import deque
+from datetime import datetime
+from enum import Enum
+from typing import Deque, Mapping, Optional, cast
+
+from arroyo import Message
+from arroyo.processing.strategies import ProcessingStrategy
+
+from snuba.subscriptions.utils import Tick
+from snuba.utils.metrics import MetricsBackend
+
+
+class SchedulerMode(Enum):
+    IMMEDIATE = "immediate"
+    WAIT_FOR_SLOWEST_PARTITION = "wait-for-slowest"
+
+
+class TickBuffer(ProcessingStrategy[Tick]):
+    """
+    The TickBuffer buffers ticks until they are ready to be submitted to
+    the next processing step.
+
+    The behavior of the TickBuffer depends on which of the two scheduler
+    modes applies.
+
+    If the scheduler mode is IMMEDIATE then there is no buffering and a
+    message is always immediately submitted to the next processing step.
+
+    If the scheduler mode is WAIT_FOR_SLOWEST_PARTITION then messages are
+    buffered until all partitions are at least up to the timestamp of the
+    tick.
+
+    `max_ticks_buffered_per_partition` applies if the scheduler mode is
+    WAIT_FOR_SLOWEST_PARTITION. Once the maximum ticks is received for that
+    partition, we start to submit ticks for processing even if that timestamp
+    is not received for all partitions yet.
+
+    This prevents the buffered tick list growing infinitely and scheduling
+    to grind to a halt if one partition starts falling far behind for some reason.
+    """
+
+    def __init__(
+        self,
+        mode: SchedulerMode,
+        partitions: int,
+        max_ticks_buffered_per_partition: Optional[int],
+        next_step: ProcessingStrategy[Tick],
+        metrics: MetricsBackend,
+    ) -> None:
+        if mode == SchedulerMode.WAIT_FOR_SLOWEST_PARTITION:
+            assert max_ticks_buffered_per_partition is not None
+
+        self.__mode = mode
+        self.__partitions = partitions
+        self.__max_ticks_buffered_per_partition = max_ticks_buffered_per_partition
+        self.__next_step = next_step
+        self.__metrics = metrics
+
+        self.__buffers: Mapping[int, Deque[Message[Tick]]] = {
+            index: deque() for index in range(self.__partitions)
+        }
+
+        # Stores the latest timestamp we received for any partition. This is
+        # just for recording the partition lag.
+        self.__latest_ts: Optional[datetime] = None
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[Tick]) -> None:
+        # If the scheduler mode is immediate or there is only one partition
+        # immediately submit message to the next step.
+        # We don't keep any latest_ts values as it is not relevant.
+        if self.__mode == SchedulerMode.IMMEDIATE or self.__partitions == 1:
+            self.__next_step.submit(message)
+            return
```

Comment:
Why not having the SchedulerMode.IMMEDIATE as a separate step (or even no step at all) and pick it when we build the strategy factory at initialization? It seems there is no reason for fitting both strategies (very different responsibilities) in the same class when the processingStrategy is already designed to make composition easy.